python/lsst/ts/mtdome/power_management/slip_ring.py

```diff
@@ -0,0 +1,246 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["SlipRing"]
+
+import logging
+
+from lsst.ts import utils
+
+from ..enums import SlipRingState
+from .power_draw_constants import (
+    CONTINUOUS_SLIP_RING_POWER_CAPACITY,
+    MAXIMUM_COOL_DOWN_TIME,
+    MAXIMUM_OVER_LOW_LIMIT_TIME,
+    MAXIMUM_SLIP_RING_POWER_CAPACITY,
+    OVER_LIMIT_POWER_AVAILABLE,
+)
+
+
+class SlipRing:
+    """A state machine implementation representing a slip ring.
+
+    The slip ring of the MTDome can handle continuous power up to a limit of 78
+    kW and temporary power up to 100 kW for up to 6 minutes. After going over
+    the continuous limit, it needs to cool down for up 4 minutes.
+
+    The exact cooling down model is unknown so a model proportional to the
+    maximum drawn power and the amlunt of time over the continuous limit is
+    used.
+
+    Parameters
+    ----------
+    log : `logging.Logger`
+        The logger for which to create a child logger.
+    index : `int`
+        For now the MTDome only has one slip ring but a second one will be
+        installed in the future.
+    disabled : `bool`
+        Is the slip ring disabled or not? Defaults to False since the currently
+        available slip ring always is enabled.
+
+    Attributes
+    ----------
+    state : `SlipRingState`
+        The state of the slip ring used in the state machine.
+    below_limit_start_time : `float`
+        The time [UNIX tai seconds] at which the power draw dropped below the
+        continuous limit. This is set by the state machine.
+    over_limit_start_time : `float`
+        The time [UNIX tai seconds] at which the power draw exceeded the
+        continuous limit. This is set by the state machine.
+    time_over_limit : `float`
+        The time interval [UNIX tai seconds] during which the power draw
+        exceeded continuous limit. This is set by the state machine.
+    cooling_down_time_needed : `float`
+        The time interval [UNIX tai seconds] for which the slip ring needs to
+        cool down. This is set by the state machine.
+    max_power_drawn : `float`
+        The maximum power drawn [W] during the time where the power draw
+        exceeded the continuous limit. This is set by the state machine.
+    """
+
+    def __init__(
+        self,
+        log: logging.Logger,
+        index: int,
+        disabled: bool = False,
+    ) -> None:
+        self.log = log.getChild(type(self).__name__)
+        self.index = index
+        self.disabled = disabled
+
+        self.state = SlipRingState.BELOW_LOW_LIMIT
+
+        self.below_limit_start_time = 0.0
+        self.over_limit_start_time = 0.0
+        self.time_over_limit = 0.0
+        self.cooling_down_time_needed = 0.0
+
+        self.max_power_drawn = 0.0
+
+    async def get_cool_down_time(self) -> float:
```

Comment:
Is there a particular reason these methods are all coroutines? Doesn't look like any of them need to be. Are you trying to future proof this against potentially having to wait for data to compute these quantities and/or do state transitions?

Reply:
No reason at all. I made them into normal methods instead.